### PR TITLE
h2o security implementation

### LINKED ISSUE #2291: H2O Security Implementation
We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner. Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of article
Your title should be descriptive of the article/tutorial. Be Specific. *Use keyword research to gain your article higher ranking.*
H2O Security Implementation

### Introduction paragraph (2-3 paragraphs):
Please write the Introductory paragraph(s), that would be included in your article. We will use this writing snippet to help us assess overall quality before approval. We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.
Take your time and write the content as you would intend to get it published. 

Security in the data analytical space is important to consider internally and externally. While surrounding a network with cybersecurity protection increases stability, the kernel within the coding program remains vulnerable. Cyber threats and attacks are unpredictable, however, this tutorial shows one of several security precautions to consider prior to starting a coding program. In this tutorial, a module in Python Programming called H<sub>2</sub>O, Internet Protocol (IP) and Domain Name System (DNS) components, and a Linux container are introduced. 

#### Prerequisites
* Linux computer Operating System (OS) or Linux container.
* Basic Bash knowledge.
* Python (preferably 3.3+, tutorial uses 3.9).
* Bash installations: H<sub>2</sub>O, Java, keytools, and a text editor.
* Python installation: H<sub>2</sub>O.
* Internet access with internet browser.

This tutorial is written at root.  

First, open a Linux container.  
[Ubuntu LTS](https://www.zdnet.com/article/minimal-ubuntu-for-containers-and-clouds/#:~:text=While%20that%20release%20is%20useful%20mostly%20for%20hobbyists%2C,the%20standard%20Docker%20Hub%20Ubuntu%2018.04%20LTS%20image.) and [Kali Linux](https://kali.org) are examples of containers.  

For any security mechanism to become stable, generation of authentication files are necessary. Refer to the instructions at the [H<sub>2</sub>O documentation website](http://docs.h2o.ai/h2o/latest-stable/h2o-docs/security.html#keystore-truststore-generation) to generate `keystore` and `truststore` authentication files.  

### Key takeaways: 
What are the 3-5 most important things the reader should understand or be able to do after reading this article? Use this area to get your ideas down on the bulk of your article or tutorial.

* Increasing security can start with the kernel.
* H<sub>2</sub>O allows security options.
* Linux systems and Linux containers can include security and protection.

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

[H<sub>2</sub>O documentation website](http://docs.h2o.ai/h2o/latest-stable/h2o-docs/security.html#keystore-truststore-generation).  
[Linux Container](https://kali.org)
[ZDnet-Ubuntu](https://www.zdnet.com/article/minimal-ubuntu-for-containers-and-clouds/#:~:text=While%20that%20release%20is%20useful%20mostly%20for%20hobbyists%2C,the%20standard%20Docker%20Hub%20Ubuntu%2018.04%20LTS%20image.)

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by pkalynan**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

This fixes #2291.

### CODE REVIEW

File: content/articles/h2o-security-implementation/index.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+#### Table of Contents
+
+* [Prerequisites](#prerequisites)
+* [Installation](#installation)
+* [Identify Change](#identify-change)
+* [Allocating Data to Lower Risk](#allocating-data-to-lower-risk)
+
+#### Prerequisites
+
+* Linux computer Operating System (OS) or Linux container with `root` authorization (this tutorial uses [Kali Linux](https://kali.org/).
```
Comment (adrianmurage) [inline content/articles/h2o-security-implementation/index.md:16]:
```suggestion
* Linux computer Operating System (OS) or Linux container with `root` authorization (this tutorial uses [Kali Linux](https://kali.org/)).
```

**Review by adrianmurage — CHANGES_REQUESTED** [review]
@pkalynan 
Going over this and reading up on H20, I realize that you actually don't implement H20 security at any one point in this tutorial.

Correct me if I'm wrong but, I would expect you to show us how to setup an instance of H20 or point us to an article that does that. Then using this as a basis take us through implementing security for it. 

Instead, you briefly touch on the security aspect, point us to the home page of H20's documentation and then move on to other ways of securing data that feel unrelated to H20.

To move forward, I need you to do one of two things:

1. Help me understand how the article covers the intended topic; "H20 security implementation"

or

2. Rewrite the article to cover the intended topic.

Kindly reply with the way you wish to move forward with this.

**Comment by pkalynan** [comment]
> @pkalynan
> Going over this and reading up on H20, I realize that you actually don't implement H20 security at any one point in this tutorial.
> 
> Correct me if I'm wrong but, I would expect you to show us how to setup an instance of H20 or point us to an article that does that. Then using this as a basis take us through implementing security for it.
> 
> Instead, you briefly touch on the security aspect, point us to the home page of H20's documentation and then move on to other ways of securing data that feel unrelated to H20.
> 
> To move forward, I need you to do one of two things:
> 
> 1. Help me understand how the article covers the intended topic; "H20 security implementation"
> 
> or
> 
> 1. Rewrite the article to cover the intended topic.
> 
> Kindly reply with the way you wish to move forward with this.

@adrianmurage 

I would like to move forward with:
1. Help me understand how the article covers the intended topic; "H20 security implementation".

Instead of restating what is already on official documentation which is what @hectorkambow advises not to do, I added a link to the tutorial as a reminder. I could redirect the reader to a specific section of the installation and start-up of h2o.

The initial steps resolve internal security within the h2o instance, so that is a type of security attempt. This is something that is not an option in other modules, and it is unique.

H2o is a module that deals with data and to increase security within h2o; data files should be secure as well. Permissions is a topic already covered in other articles, so it was just there as a reminder.

Another is extremely backend solution if someone connected h2o to a Java database, where a basic Java If-then statement is an attempt to prevent SQL injection attacks. This is to help secure data from user input if an attack was targeting the database connected to h2o. Java is not too irrelevant since databases can be written in Java. Majority of data files came from a database at one point in the data analysis cycle which is why database was included.

I could change the title to become more appropriate. Options could include, "Data protection", or "H2O Data Protection". I could also explain in more detail what h2o is used for in this tutorial which is to perform data analyses.

**Comment by adrianmurage** [comment]
@pkalynan 
I agree with your approach.

Directing the reader to the specific sections in the documentation would be really helpful.

I think you should consider giving some information about H20 from the perspective of the 5W and H
(What, Why, Where, When, What, How). 
Like the idea that H20 is unique(and why its unique) is a good example of such an elaboration.

I like the idea to align the Title with the content you intend to write about.

All in all, yes I now think I have an understanding of what your intention was.
Kindly @ me when you've updated the document.

Great job with the explanation to me. I appreciate it.

**Comment by pkalynan** [comment]
@adrianmurage 

I made some changes as per your recommendations.

If you could let me know a title you would prefer, I can change the hero image and index.md file accordingly. 

Thank you in advance.

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by adrianmurage** [comment]
> 
> 
> @adrianmurage
> 
> I made some changes as per your recommendations.
> 
> If you could let me know a title you would prefer, I can change the hero image and index.md file accordingly.
> 
> Thank you in advance.

A heading that mentions the aspect of securing the data like: "Securing data with h2o" would be appropriate.
This is just a suggestion, feel free to tweak it as you see fit.

**Review by adrianmurage — CHANGES_REQUESTED** [review]
@pkalynan 
I am done with the review.
Great job on this one! Looking forward to seeing more of your work.

Kindly finalize with the title of the article to enable me to update the front matter and label this ready for final review

@ me when you're done.

**Comment by pkalynan** [comment]
> @pkalynan
> I am done with the review.
> Great job on this one! Looking forward to seeing more of your work.
> 
> Kindly finalize with the title of the article to enable me to update the front matter and label this ready for final review
> 
> @ me when you're done.

@adrianmurage,

Title changed in response to your helpful suggestion.

Thank you.

**Comment by adrianmurage** [comment]
> 
> 
> > @pkalynan
> > I am done with the review.
> > Great job on this one! Looking forward to seeing more of your work.
> > Kindly finalize with the title of the article to enable me to update the front matter and label this ready for final review
> > @ me when you're done.
> 
> @adrianmurage,
> 
> Title changed in response to your helpful suggestion.
> 
> Thank you.

The PR has two `index.md` files.
Kindly make sure you delete the old `index.md` file and its directory `h2o-security-implementation`

**Comment by adrianmurage** [comment]
@pkalynan 
Great job with this article!
I'll make final changes then mark it for final review
Looking forward to more of your work

**Review by adrianmurage — APPROVED** [review]

**Review by adrianmurage — APPROVED** [review]

**Comment by hectorkambow** [comment]
great work @pkalynan 

Very easy to follow along 

cc @adrianmurage  🚀

**Pull request merged**